components/builder-depot/src/server.rs

```diff
@@ -2307,6 +2264,393 @@ fn target_from_headers(user_agent_header: &UserAgent) -> result::Result<PackageT
+                }
+            } else {
+                if channel.is_some() {
+                    info!("Installing packages from an upstream depot and the channel wasn't stable. Instead, it was {}", channel.unwrap());
+                }
+
+                match OriginPackage::from_archive(&mut archive) {
+                    Ok(p) => Ok(p),
+                    Err(e) => Err(Error::HabitatCore(e)),
+                }
+            }
+        }
+        Err(e) => {
+            warn!(
+                "Failed to download {} from upstream_depot {:?}. e = {:?}",
+                ident, upstream_depot, e
+            );
+            Err(Error::DepotClientError(e))
+        }
+    }
+}
+
+fn check_package_from_upstream_depot(
+    depot: DepotUtil,
+    depot_cli: DepotClient,
+    oauth_token: Option<String>,
+    upstream_depot: String,
+    ident: OriginPackageIdent,
+    channel: Option<String>,
+) -> Result<OriginPackage> {
+    let another_token = oauth_token.clone();
+    let token_ref = another_token.as_ref().map(|x| &**x);
+    let another_channel = channel.clone();
+    let channel_ref = another_channel.as_ref().map(|x| &**x);
+
+    match depot_cli.show_package(&ident, channel_ref, token_ref) {
+        Ok(mut package) => {
+            let pkg_ident: PackageIdent = ident.clone().into();
+            let remote_pkg_ident: PackageIdent = package.take_ident().into();
+
+            if remote_pkg_ident > pkg_ident {
+                let opi: OriginPackageIdent = OriginPackageIdent::from(remote_pkg_ident);
+                download_package_from_upstream_depot(
+                    depot,
+                    depot_cli,
+                    oauth_token,
+                    &upstream_depot,
+                    opi,
+                    channel,
+                )
+            } else {
+                Err(Error::IronResponse(Response::with(status::NotFound)))
+            }
+        }
+        Err(e) => {
+            debug!(
+                "Failed to show package {} from {}: {:?}",
+                ident, upstream_depot, e
+            );
+            Err(Error::DepotClientError(e))
+        }
+    }
+}
+
+fn refresh_package_from_upstream_depot(
+    req: &mut Request,
+    ident: OriginPackageIdent,
+    channel: Option<String>,
+) {
+    let lock = req.get::<persistent::State<DepotUtil>>()
+        .expect("depot not found");
+    let d = lock.read().expect("depot read lock is poisoned");
+    let depot = d.clone();
+    let config = d.config.clone();
+
+    if config.upstream_depot.is_none() {
+        return;
+    }
+
+    let depot_cli = depot_client_from_config(&config).unwrap();
+    let upstream_depot = config.upstream_depot.unwrap(); // unwrap is fine because we just checked if it was None
+    let oauth_token = helpers::get_optional_oauth_token(req);
+
+    // Let's do the actual work that takes time in the background so the main request handler can
+    // continue on its way.
+    spawn(move || {
```

Comment:
Instead of spawning new threads for requests, we may want to consider having a long lived background thread (similar to how we are handling worker manager and scheduler threads) that is responsible for pulling refresh tasks from a list of idents.  The main thread then just has to add an ident to the refresh list (if it's not already present), and let the worker manage churning thru that list every N second intervals.  The main thread could also add multiple idents to refresh (eg, any updated dependencies for the package that is being fetched could all be added at once).

Reply:
I think this is a great idea, but it feels like a pretty big change to make at this point in the PR, for what is essentially a nice-to-have.  I will make a note to file an issue for this kind of enhancement after this PR merges.

Reply:
Sounds good - should be an easy copy-pasta from other of the other modules